src/Persistence_matrix/include/gudhi/Persistence_matrix/columns/heap_column.h

```diff
@@ -0,0 +1,1082 @@
+  Heap_column(const Container_type& nonZeroRowIndices, Column_settings* colSettings);
+  template <class Container_type = typename Master_matrix::boundary_type>
+  Heap_column(const Container_type& nonZeroChainRowIndices, 
+              dimension_type dimension, 
+              Column_settings* colSettings);
+  Heap_column(const Heap_column& column, Column_settings* colSettings = nullptr);
+  Heap_column(Heap_column&& column) noexcept;
+  ~Heap_column();
+
+  // just for the sake of the interface
+  // row containers and column index are ignored as row access is not implemented for heap columns
+  template <class Container_type = typename Master_matrix::boundary_type, class Row_container_type>
+  Heap_column(index columnIndex, 
+              const Container_type& nonZeroRowIndices, 
+              Row_container_type* rowContainer,
+              Column_settings* colSettings);
+  template <class Container_type = typename Master_matrix::boundary_type, class Row_container_type>
+  Heap_column(index columnIndex, 
+              const Container_type& nonZeroChainRowIndices, 
+              dimension_type dimension,
+              Row_container_type* rowContainer, 
+              Column_settings* colSettings);
+  template <class Row_container_type>
+  Heap_column(const Heap_column& column, 
+              index columnIndex, 
+              Row_container_type* rowContainer,
+              Column_settings* colSettings = nullptr);
+
+  std::vector<Field_element_type> get_content(int columnLength = -1) const;
+  bool is_non_zero(id_index rowIndex) const;
+  bool is_empty();
+  std::size_t size() const;
+
+  template <class Map_type>
+  void reorder(const Map_type& valueMap, [[maybe_unused]] index columnIndex = -1);
+  void clear();
+  void clear(id_index rowIndex);
+
+  id_index get_pivot();
+  Field_element_type get_pivot_value();
+
+  iterator begin() noexcept;
+  const_iterator begin() const noexcept;
+  iterator end() noexcept;
+  const_iterator end() const noexcept;
+  reverse_iterator rbegin() noexcept;
+  const_reverse_iterator rbegin() const noexcept;
+  reverse_iterator rend() noexcept;
+  const_reverse_iterator rend() const noexcept;
+
+  template <class Cell_range>
+  Heap_column& operator+=(const Cell_range& column);
+  Heap_column& operator+=(Heap_column& column);
+
+  Heap_column& operator*=(unsigned int v);
+
+  // this = v * this + column
+  template <class Cell_range>
+  Heap_column& multiply_and_add(const Field_element_type& val, const Cell_range& column);
+  Heap_column& multiply_and_add(const Field_element_type& val, Heap_column& column);
+  // this = this + column * v
+  template <class Cell_range>
+  Heap_column& multiply_and_add(const Cell_range& column, const Field_element_type& val);
+  Heap_column& multiply_and_add(Heap_column& column, const Field_element_type& val);
+
+  friend bool operator==(const Heap_column& c1, const Heap_column& c2) {
+    if (&c1 == &c2) return true;
+    return c1.get_content() == c2.get_content();
```

Comment:
If the column is relatively sparse (most common case?), going through a dense type is going to be costly. Wouldn't it be cheaper to remove redundancies (`_prune`?), sort (already done by `_prune`?) and compare? You could even try to do it progressively, check if the 2 columns have the same pivot and recurse, but that may be inconvenient to do non-destructively.
(similar comment for all calls to get_content in this file)

Reply:
True, if the pivot is big, the comparison will be uselessly costly. The problem is that the two columns has to be `const` so I cannot directly prune them. On the other side, copying them is probably still cheaper for the column containing high values. And once copied, there is no problem with destroying the column with `_pop_pivot`s.

Reply:
It could make sense to mark `_prune` as `const` (and add `mutable` where needed for it to work). The column is logically the same after pruning. The main issues with that are
* iterators allow users to see the arbitrary internal order and are invalidated by pruning
* if we ever want to use a column in multiple threads, `const` is often taken to mean "safe to read from multiple threads at the same time"